File: 0800-letter-case-permutation/0800-letter-case-permutation.java
class Solution {
    List<String> ans = new ArrayList<>();
    public void helper (int index, StringBuilder s) {
        // BASE CASE
        if (index == s.length()) {
            ans.add(s.toString());
            return;
        }
        
        char ch = s.charAt(index);

        // RECURSION
        if (Character.isLetter(ch)) {
            // Lowercase branch
            s.setCharAt(index, Character.toLowerCase(ch));
            helper (index + 1, s);
            // Uppercase branch
            s.setCharAt(index, Character.toUpperCase(ch));
            helper(index + 1, s);
        } else {
            helper (index + 1, s);
        }
    }
    public List<String> letterCasePermutation(String s) {
        helper(0, new StringBuilder(s));
        return ans;
    }
}

File: 0800-letter-case-permutation/0800-letter-case-permutation.cpp
class Solution {
public:
    vector<string> str;
    int n;
    void helper(int index, string s) {
        // BASE CASE
        if (index == s.size()) {
            str.push_back(s);
            return;
        }
        // RECURSION
        if (isalpha(s[index])) {
            s[index] = tolower(s[index]); // Lowercase branch
            helper(index + 1, s);
            s[index] = toupper(s[index]); // Uppercase branch
            helper(index + 1, s);
        } else {
            helper
            (index + 1, s);
        }
    }
    vector<string> letterCasePermutation(string s) {
        helper(0, s);
        return str;
    }
};